mtags/src/main/scala-3/scala/meta/internal/pc/SemanticTokenProvider.scala

```diff
@@ -0,0 +1,195 @@
+package scala.meta.internal.pc
+
+import scala.meta.internal.mtags.MtagsEnrichments.*
+import scala.meta.internal.parsers.SoftKeywords
+import scala.meta.pc.VirtualFileParams
+import scala.meta.tokens.*
+
+import dotty.tools.dotc.ast.tpd.*
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Flags
+import dotty.tools.dotc.core.Symbols.Symbol
+import dotty.tools.dotc.interactive.InteractiveDriver
+import dotty.tools.dotc.util.SourcePosition
+import org.eclipse.lsp4j.SemanticTokenModifiers
+import org.eclipse.lsp4j.SemanticTokenTypes
+
+/**
+ *  Provides semantic tokens of file(@param params)
+ *  according to the LSP specification.
+ */
+final class SemanticTokenProvider(
+    driver: InteractiveDriver,
+    params: VirtualFileParams,
+) extends SemTokenProvider:
+  type Sym = Symbol
+  type Pos = SourcePosition
+
+  implicit val ord: Ordering[NodeInfo] = Ordering.fromLessThan((ni1, ni2) =>
+    if ni1.pos.start == ni2.pos.start then ni1.pos.end < ni2.pos.end
+    else ni1.pos.start < ni2.pos.start
+  )
+  override def getTokens(): Tokens =
+    import scala.meta.*
+    implicit val dialect = scala.meta.dialects.Scala3
+    params.text().tokenize.get
+
+  object Collector extends PcCollector[NodeInfo](driver, params):
+    override def collect(
+        parent: Option[Tree]
+    )(tree: Tree, pos: SourcePosition, symbol: Option[Symbol]): NodeInfo =
+      val sym = symbol.fold(tree.symbol)(s => s)
+      if pos.exists then NodeInfo(sym, adjust(pos)._1)
+      else NodeInfo(sym, pos)
+  given Context = Collector.ctx
+  val nodes0: List[NodeInfo] =
+    Collector.result()
+  override val nodes = nodes0.filter(_.pos.exists).sorted
+
+  /**
+   * The position of @param tk must be incremented from the previous call.
+   */
+  def pickFromTraversed(
+      tk: scala.meta.tokens.Token,
+      nodesIterator: List[NodeInfo],
+  )(using Context): Option[(NodeInfo, List[NodeInfo])] =
+
+    def isTarget(node: NodeInfo): Boolean =
+      node.pos.start == tk.pos.start &&
+        node.pos.end == tk.pos.end
+    val candidates = nodesIterator.dropWhile(_.pos.start < tk.pos.start)
+
+    def filterOutOwner(sym: Symbol): Boolean =
+      sym.owner.isPrimaryConstructor || sym.owner.decodedName == "copy"
+        || sym.owner.decodedName == "apply"
+
+    def filterOutFlags(sym: Symbol): Boolean =
+      sym.is(Flags.Synthetic) ||
+        (sym.is(Flags.Implicit) && sym.is(Flags.Method)) ||
+        (sym.owner.is(Flags.Implicit) && sym.owner.is(Flags.Method))
+
+    def classOverMethod(nodes: List[NodeInfo]): Option[NodeInfo] =
+      nodes.find { case ni => !ni.sym.is(Flags.Method) } match
+        case Some(node) => Some(node)
+        case None => nodes.headOption
+
+    candidates
+      .takeWhile(_.pos.start == tk.pos.start)
+      .filter(isTarget) match
```

Comment:
We can't do that, consider case there token has start, end `(2,6)` and nodes have `(2,4)`, `(2,6)`, `(2,8)`, using this method we would fail of `(2,4)` and not find `(2,6)`

Reply:
ach ok, that's fine then. I don't think we have another standard libary emthod that fits here.